### PR TITLE
Blog app - integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by RashidMenhas**
## In this milestone I achieved :100: 

- [ ] Write integration tests with Capybara gem.

- [ ] Secure app from n+1 problems.

#### :diamond_shape_with_a_dot_inside: Integration specs
- Use Capybara to write integration tests for each view in your project.

- Think on the flow of things and group the tests of page in a way that makes sense. For example, one test would check for the things in the page and another test for clicking links on it.

###  :diamond_shape_with_a_dot_inside: User index page:

- I can see the username of all other users.
- I can see the profile picture for each user.
- I can see the number of posts each user has written.
- When I click on a user, I am redirected to that user's show page.

###  :diamond_shape_with_a_dot_inside: user show page:

- I can see the user's profile picture.
- I can see the user's username.
- I can see the number of posts the user has written.
- I can see the user's bio.
- I can see the user's first 3 posts.
- I can see a button that lets me view all of a user's posts.
- When I click a user's post, it redirects me to that post's show page.
- When I click to see all posts, it redirects me to the user's post's index page.

###  :diamond_shape_with_a_dot_inside: User post index page:

- I can see the user's profile picture.
- I can see the user's username.
- I can see the number of posts the user has written.
- I can see a post's title.
- I can see some of the post's body.
- I can see the first comments on a post.
- I can see how many comments a post has.
- I can see how many likes a post has.
- I can see a section for pagination if there are more posts than fit on the view.
- When I click on a post, it redirects me to that post's show page.
 
###  :diamond_shape_with_a_dot_inside: Post show page:

- I can see the post's title.
- I can see who wrote the post.
- I can see how many likes it has.

- I can see the post body.
- I can see the username of each commentor.
- I can see the comment each commentor left.

Screenshot from 2023-03-02 23-20-45

Screenshot from 2023-03-02 23-23-50

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -178,8 +179,9 @@ You can deploy this project using:
 
 > Describe 1 - 3 features you will add to the project.
```
Comment (TadesseDev) [inline README.md]: - [ ] Please notice the section in your readme which start with `>` are directions/guidelines from microverse to help you understand what to write in that section and should not be part of your readme file. Please consider removing this and all the like contents from the readme file.

File: README.md [hunk 2]
```diff
@@ -95,8 +95,9 @@ After you're finished please remove all the comments and instructions!
 ### Key Features <a name="key-features"></a>
 
 > Describe between 1-3 key features of the application.
-- This is Rails based project.
```
Comment (TadesseDev) [inline README.md]: - [ ] Please notice the section in your readme which start with `>` are directions/guidelines from microverse to help you understand what to write in that section and should not be part of your readme file. Please consider removing this and all the like contents from the readme file.

**Review by TadesseDev — CHANGES_REQUESTED** [review]
### Hi @RashidMenhas, 

### I'm Tadesse Alemayehu your reviewer for this PR. :hand: 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request 🛑🛑
yPkfDFf

- [ ] Please consider removing non-important contents from your readme, The template is designed by Microverse for us all to use to document our projects professionally and has it filled with content related to each project we may be working on. currently, I can see you have not removed the guiding contents  which start with `>` and `directions set by Microverser`

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

File: spec/views/users/show_spec.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  end
+  it 'Shows Author biography' do
+    expect(page.body).to have_content(@author.bio)
+  end
+
+  it 'Shows the Author photo' do
+    image = page.all('img')
+    expect(image.size).to eq 1
+  end
+  it 'Shows number of posts Author has written' do
```
Comment (Davidosky007) [inline spec/views/users/show_spec.rb:33]: * Kindly add Tests that assert you can see the user's first 3 posts as its missing 🙏

File: spec/views/posts/index_spec.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    visit user_path(@author)
+
+    click_on @author.name
+    click_link 'See all posts'
+  end
+
+  it 'Shows the username of the user' do
+    expect(page).to have_content(@author.name)
+  end
```
Comment (Davidosky007) [inline spec/views/posts/index_spec.rb:25]: * Kindly add Tests that assert you can see a section for pagination if there are more posts than fit on the view and 
* Tests that assert you can see the first comments on a post as these tests are missing

File: spec/views/posts/show_spec.rb [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    @com2 = Comment.create(author: @author, post: @post,
+                           text: 'Hello there Momanyi Hassan, this Dero I am watching you')
+    @com3 = Comment.create(author: @author, post: @post, text: 'Spend the credits wisely :|')
+
+    @like1 = @post.likes.create!(author: @author)
+
+    visit user_post_path(@author, @post)
+  end
+
+  it 'Shows the username of the author' do
```
Comment (Davidosky007) [inline spec/views/posts/show_spec.rb]: * Kindly add Tests that assert you can see who wrote the post as this test is missing 🙏

**Review by Davidosky007 — CHANGES_REQUESTED** [review]
Hi @RashidMenhas @demix007 👋, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights

* No linter errors 👍 
* All tests passed ✔️ 
* Good looking readme file 👍 
## Required Changes ♻️
* As stated in the [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#:~:text=In%20your%20PR%20description%20include%20the%20screenshots%20of%20your%20app%27s%20logs%20before%20and%20after%20the%20fix) kindly add In your PR description the  screenshots of your app's logs before and after the fix
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @davidsoky007 in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/views/posts/show_spec.rb (hunk 5 above)
Comment (rloterh) [reply spec/views/posts/show_spec.rb]: resolved! ✔️

File: spec/views/users/show_spec.rb [hunk 6]
```diff
@@ -15,13 +15,19 @@
                          text: 'Do more practice if your want to success in your life.')
     visit user_path(@author)
   end
-  it 'Shows Athor name' do
+  it 'Shows Author name' do
     expect(page).to have_content(@author.name)
   end
   it 'Shows Author biography' do
     expect(page.body).to have_content(@author.bio)
   end
```
Comment (rloterh) [inline spec/views/users/show_spec.rb]: - [ ] _Tests that asserts you can see the user's first 3 posts_ is still missing. Let's include this missing test case.

File: spec/views/posts/index_spec.rb (hunk 4 above)
Comment (rloterh) [reply spec/views/posts/index_spec.rb:25]: resolved! ✔️

File: spec/views/users/show_spec.rb (hunk 3 above)
Comment (rloterh) [reply spec/views/users/show_spec.rb:33]: Needs attention!

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Rashid, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project but you are almost there! 🚦 🚦   🏃 

## Highlights 👍  
- Good job so far with putting up integration specs!
- ReadMe looks good!

## Required Changes ♻️
- [ ] Please be reminded that in this activity we want you to add screenshots of the logs before and after the `N +1` fix to the Pull Request description, as indicated [here](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md).

_Kindly check the other inline comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

**You can also consider:**
- MIT license link in your ReadMe is broken, kindly endeavor to fix it.

Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/views/users/show_spec.rb (hunk 3 above)
Comment (rloterh) [reply spec/views/users/show_spec.rb:33]: resolved! ✔️

**Review by rloterh — APPROVED** [review]
## Status: Approved :heavy_check_mark:

Hi Rashid,  

Your project is complete! 🏁 There is nothing else to say other than... it's time to merge it 🍾 Congratulations!  🎉🌟 

## Highlights 👍
- Good job implementing recommended changes!

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code._

**You can also consider:**
- Kindly endeavor to remove the guide note below the MIT license.

Cheers..Happy coding! 💻  🍷  🚀 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**